common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,905 @@
+     */
+    void updateChunkInfo(long chunkOffset, int mapIndex) throws IOException {
+      long idxStartPos = -1;
+      try {
+        // update the chunk tracker to meta file before index file
+        writeChunkTracker(mapIndex);
+        idxStartPos = indexChannel.position();
+        logger.trace("{} shuffleId {} reduceId {} updated index current {} updated {}",
+          appShuffleId.appId, appShuffleId.shuffleId, reduceId, this.lastChunkOffset,
+          chunkOffset);
+        indexWriteStream.writeLong(chunkOffset);
+      } catch (IOException ioe) {
+        if (idxStartPos != -1) {
+          // reset the position to avoid corrupting index files during exception.
+          logger.warn("{} shuffleId {} reduceId {} reset index to position {}",
+            appShuffleId.appId, appShuffleId.shuffleId, reduceId, idxStartPos);
+          indexChannel.position(idxStartPos);
+        }
+        throw ioe;
+      }
+      this.lastChunkOffset = chunkOffset;
+    }
+
+    private void writeChunkTracker(int mapIndex) throws IOException {
+      if (mapIndex == -1) {
+        return;
+      }
+      chunkTracker.add(mapIndex);
+      if (trackerBuf == null) {
+        trackerBuf = Unpooled.buffer(Encoders.Bitmaps.encodedLength(chunkTracker));
+      }
+      Encoders.Bitmaps.encode(trackerBuf, chunkTracker);
+      long metaStartPos = metaChannel.position();
+      try {
+        logger.trace("{} shuffleId {} reduceId {} mapIndex {} write chunk to meta file",
+          appShuffleId.appId, appShuffleId.shuffleId, reduceId, mapIndex);
+        metaChannel.write(trackerBuf.nioBuffer());
+      } catch (IOException ioe) {
+        logger.warn("{} shuffleId {} reduceId {} mapIndex {} reset position of meta file to {}",
+          appShuffleId.appId, appShuffleId.shuffleId, reduceId, mapIndex, metaStartPos);
+        metaChannel.position(metaStartPos);
+        throw ioe;
+      } finally {
+        trackerBuf.clear();
+      }
+    }
+
+    void closeAllFiles() throws IOException {
+      this.dataChannel.close();
+      this.metaChannel.close();
+      this.indexWriteStream.close();
+      this.indexChannel.close();
+    }
+  }
+
+  /**
+   * Wraps all the information related to the merge directory of an application.
+   */
+  private static class AppPathsInfo {
+
+    private final String user;
+    private String[] activeLocalDirs;
+
+    AppPathsInfo(String user) {
+      this.user = Preconditions.checkNotNull(user, "user cannot be null");
+    }
+
+    private AppPathsInfo updateActiveLocalDirs(
+        String appId, String relativePathPattern, String[] localDirs) {
+      if (activeLocalDirs == null) {
+        String relativePath = String.format(relativePathPattern, user, appId);
+        activeLocalDirs = Arrays.stream(localDirs)
+          .map(localDir -> localDir.substring(0, localDir.indexOf(relativePath)))
+          .toArray(String[]::new);
+        logger.info("Updated the active local dirs {} for application {}",
```

Comment:
nit: Move this within `logger.isInfoEnabled` given the `Arrays.toString` in the parameter.

Reply:
done